Included initial exploratory analysis notebook (#794)
AlexsLemonade/OpenScPCA-analysis: analyses/cell-type-dsrct/exploratory_analysis/01-marker-gene-tumor-classification.Rmd

for(i in 1:length(marker_sum_exp)) {
  p <- ggplot(marker_sum_exp[[i]], aes(x = UMAP1, y = UMAP2, color = sum_classification)) +
  geom_point(size = 0.5, alpha = 1) +
  ggtitle(names(umap_df_list_excluded_scaled)[i])
  print(p)
}

```

Based on my experiences, the above plots show that we are unable to adequately determine the tumor cells from the normal cells. This can be seen in SCPCS000729, which is most likely all tumor cells, since it is collected from a patient-derived xenograft.

## Conclusions 

- We do see variation in DSRCT-specific gene expression across cells, suggesting that there may be gene drop-out reducing the quality of the data. 
- `ST6GALNAC5`, `PTRPQ`, and `IQCJ-SCHIP1` are good markers for the DSRCT cells within this data. The other markers are not as highly expressed either due to heterogeneity or data quality. 
- For the next steps, we may try to use `SingleR` or `CellAssign` to identify the normal cells and then work to identify remaining cells as tumor cells. Normal cells may have more definitive markers than tumor cells. 

## Save outputs 

```{r}
# get an RDS of the processed data

> saveRDS(sce_file_list, '../results/SCPCP000013_sce_file_list.rds')

Reviewer: This seems to just be the file list? Did you want to also save the extracted data tables?

Reviewer: Should we save the extracted data not jus the file names?
```suggestion
# save an RDS of the processed data
saveRDS(umap_df_list, file.path(module_base, 'results/SCPCP000013_umap_df_list.rds')
```

    dplyr::select(barcodes,
                  UMAP1,
                  UMAP2,
                  gene_symbol,
                  ensembl_gene_id,
                  gene_expression,
                  cluster)
})

names(umap_df_list) <- metadata_DSRCT$scpca_sample_id
```

```{r, fig.width=8}
for(i in 1:length(umap_df_list)){
  # faceted umap showing a umap panel for each marker gene
  p <- ggplot(umap_df_list[[i]], aes(x = UMAP1, y = UMAP2, color = gene_expression)) +
    geom_point(alpha = 0.8, size = 0.1) +
    facet_wrap(vars(gene_symbol)) +
    scale_color_viridis_c() +
    labs(color = "Log-normalized gene expression") +
    # remove axis numbers and background grid
    scale_x_continuous(labels = NULL, breaks = NULL) +
    scale_y_continuous(labels = NULL, breaks = NULL) +
    theme(
      aspect.ratio = 1,
      legend.position = "bottom",
      axis.title = element_text(size = 9, color = "black"),
      strip.text = element_text(size = 8),
      legend.title = element_text(size = 9),
      legend.text = element_text(size = 8)
    ) +
    guides(colour = guide_colorbar(title.position = "bottom", title.hjust = 0.5)) +
    ggtitle(names(umap_df_list)[i])
  

>   tiff(filename = paste0('../plots/marker_expression_', names(umap_df_list)[i], '.tiff'), width = 6, height = 6, units = 'in', res = 150)
>   print(p)
>   dev.off()

Reviewer: I'd suggest using `ggsave` to allow one-command saving, and `png` should be smaller than tiff (though this may be something to test).

```suggestion
  ggsave(
    filename = paste0('../plots/marker_expression_', names(umap_df_list)[i], '.png'), 
    plot = p,
    width = 6, 
    height = 6, 
    units = 'in', 
    dpi = 150
  )
```

Reviewer: Please also include the plot files (and the rendered html if possible) in the submission.

Author: Thanks. Will add.

```{r}
for(i in 1:length(umap_df_list_excluded)) {
  p <- ggplot(umap_df_list_excluded[[i]],
              aes(x = gene_expression, fill = gene_symbol)) +
    geom_density() +
    facet_wrap(vars(gene_symbol), scales = 'free_y') +
    theme(legend.position = "none") + 
    ggtitle(names(umap_df_list_excluded)[i])
  print(p)
  
  #save plots
  ggsave(
    filename = paste0(
      '../plots/marker_distribution_',
      names(umap_df_list)[i],
      '.png'
    ),
    plot = p,
    width = 6,
    height = 6,
    units = 'in',
    dpi = 150
  )
}
```

Although we see some slight semblance of a bimodal distribution for most marker genes, it is hard to see and we cannot directly compare gene expression values across genes. 
This would make it hard to identify a cut off to categorize cells as expression or not expressing the marker gene. 

Now we will transform each of the gene expression vectors by generating z-scores. 

> Then we might be able to find a cut off we can use across samples for if marker genes are present in a cell or not. 

Reviewer: I think this is fine for a rough look, but it is quite brittle. As you note later, some samples may contain only tumor cells. In that case, using the zscore may not be a good measure; what we really want is to find the division between the bimodal peaks. This is beyond the scope of this PR, but for your final classifications I would not expect you to want to use z-scoring.

```{r}

for(i in 1:length(umap_df_list_excluded_scaled)) {
  p <- ggplot(umap_df_list_excluded_scaled[[i]],
              aes(x = transformed_gene_expression, fill = gene_symbol)) +
    geom_density() +
    facet_wrap(vars(gene_symbol), scales = 'free_y') +
    theme(legend.position = "none") + 
    ggtitle(names(umap_df_list_excluded_scaled)[i])
  
  print(p)
  
  #save plots
  ggsave(
    filename = paste0(
      '../plots/marker_distribution_zscaled_',
      names(umap_df_list)[i],
      '.png'
    ),
    plot = p,
    width = 6,
    height = 6,
    units = 'in',
    dpi = 150
  )
  
}
```

It looks like some marker genes have distinct groups of cells with z-score > 0, while other marker genes may not be as informative. 

### Classify tumor cells using marker genes only

Let's try and use the marker gene expression to classify tumor cells. 
It looks like we could use a cutoff of z-score > 0 to count that cell as a tumor cell. 

We could either count any cell that expresses at least one marker gene > 0 as a tumor cell, or look at the combined expression. 
Let's start with classifying tumor cells as tumor if any marker gene is present (z-score > 0). 

> This also means the sum of all z-scores > 0. 

Reviewer: I am confused by this statement. Why would having one marker gene > 0 mean the sum of all is > 0?

Again, I'd probably not want to use the z scores for this, but maybe instead scaling just by range (maybe the 90% quartile or something like that?), but leaving zero as the lower bound? The cutoff may still be hard to draw, but z scoring often breaks when there are not two relatively equal classes. (I should say, one of my current projects is to find a robust way to automatically segment these kinds of distributions.)

Below, we can get the sum of the transformed gene expression of all marker genes and plot in a single UMAP. 

```{r}
# calculate sum gene expression across all marker genes in list 
marker_sum_exp <- lapply(umap_df_list_excluded_scaled, function(x){
  x |>
  dplyr::group_by(barcodes) |> 
  dplyr::mutate(sum_exp = sum(transformed_gene_expression, na.rm = T)) |> 
  dplyr::select(barcodes, UMAP1, UMAP2, sum_exp, cluster) |> 
  dplyr::distinct()
})

# plot mean gene expression 

for(i in 1:length(marker_sum_exp)) {
  p <- ggplot(marker_sum_exp[[i]], aes(x = UMAP1, y = UMAP2, color = sum_exp)) +
  geom_point(size = 0.5, alpha = 0.5) +
  scale_color_viridis_c() +
    ggtitle(names(umap_df_list_excluded_scaled)[i])
  print(p)
  
  #save plots
  ggsave(
    filename = paste0(

>       '../plots/UMAP_marker_expression_',

Reviewer: Since we have already defined the module base, we can use that to be a bit more robust to moving the notebook file.
```suggestion
      module_base,
      '/plots/UMAP_marker_expression_',
```

      names(umap_df_list)[i],
      '.png'
    ),
    plot = p,
    width = 6,
    height = 6,
    units = 'in',
    dpi = 150
  )
}

```
Similar to the individual plots, it looks like there is one group of cells on the bottom right that has the highest marker gene expression. 
We would anticipate that these are most likely to be the tumor cells. 

Now let's classify any cell that has a sum of marker genes > 0 (after z-transformation) as tumor cells. 

```{r}
# classify tumor cells based on presence of any marker genes 
marker_sum_exp <- lapply(marker_sum_exp, function(x){
  x |> 
  dplyr::mutate(sum_classification = dplyr::if_else(sum_exp > 0, "Tumor", "Normal"))})

for(i in 1:length(marker_sum_exp)) {
  p <- ggplot(marker_sum_exp[[i]], aes(x = UMAP1, y = UMAP2, color = sum_classification)) +
  geom_point(size = 0.5, alpha = 1) +
  ggtitle(names(umap_df_list_excluded_scaled)[i])
  print(p)
  
  #save plots
  ggsave(
    filename = paste0(

>       '../plots/UMAP_classification_',

Reviewer: ```suggestion
      module_base,
      '/plots/UMAP_classification_',
```

      names(umap_df_list)[i],
      '.png'
    ),
    plot = p,
    width = 6,
    height = 6,
    units = 'in',
    dpi = 150
  )
}

```

This gives us a rough idea of cells that may be classified as tumor cells. However, I believe re-doing this analysis using the cluster level data may give us better results. This is highly dependent on whether the clusters were generated with enough granularity.

```{r}
# calculate sum gene expression across all marker genes in list 
marker_sum_exp <- lapply(umap_df_list_excluded_scaled, function(x){
  x |> 
  dplyr::group_by(cluster) |> #change to cluster
  dplyr::mutate(sum_exp = sum(transformed_gene_expression, na.rm = T)) |> 
  dplyr::select(barcodes, UMAP1, UMAP2, sum_exp, cluster) |> 
  dplyr::distinct()
})

# plot mean gene expression 

for(i in 1:length(marker_sum_exp)) {
  p <- ggplot(marker_sum_exp[[i]], aes(x = UMAP1, y = UMAP2, color = sum_exp)) +
  geom_point(size = 0.5, alpha = 0.5) +
  scale_color_viridis_c() +
    ggtitle(names(umap_df_list_excluded_scaled)[i])
  print(p)
  
  #save plots
  ggsave(

>     filename = paste0('../plots/UMAP_marker_expression_cluster_', names(umap_df_list)[i], '.png'),

Reviewer: ```suggestion
    filename = paste0(module_base, '/plots/UMAP_marker_expression_cluster_', names(umap_df_list)[i], '.png'),
```

    plot = p,
    width = 6,
    height = 6,
    units = 'in',
    dpi = 150
  )
}

```
It looks like there are group of cells with high marker gene expression. We  anticipate that these are most likely to be the tumor cells. The groups with low or negative values are likely normal cells. 

Now let's classify clusters that has a sum of marker genes > 0 (after z-transformation) as tumor cell clusters. 

```{r}
# classify tumor cells based on presence of any marker genes 
marker_sum_exp <- lapply(marker_sum_exp, function(x){
  d <- density(x$sum_exp)
  
  x |> 
  dplyr::mutate(sum_classification = dplyr::if_else(sum_exp > 0, "Tumor", "Normal"))})

for(i in 1:length(marker_sum_exp)) {
  p <- ggplot(marker_sum_exp[[i]], aes(x = UMAP1, y = UMAP2, color = sum_classification)) +
  geom_point(size = 0.5, alpha = 1) +
  ggtitle(names(umap_df_list_excluded_scaled)[i])
  print(p)

  #save plots
  ggsave(
    filename = paste0(

>       '../plots/UMAP_classification_cluster_',

Reviewer: ```suggestion
      module_base,
      '../plots/UMAP_classification_cluster_',
```

  #save plots
  ggsave(
    filename = paste0(module_base, '/plots/UMAP_marker_expression_cluster_', names(umap_df_list)[i], '.png'),
    plot = p,
    width = 6,
    height = 6,
    units = 'in',
    dpi = 150
  )
}

```
It looks like there are group of cells with high marker gene expression. We  anticipate that these are most likely to be the tumor cells. The groups with low or negative values are likely normal cells. 

Now let's classify clusters that has a sum of marker genes > 0 (after z-transformation) as tumor cell clusters. 

```{r}
# classify tumor cells based on presence of any marker genes 
marker_sum_exp <- lapply(marker_sum_exp, function(x){
  d <- density(x$sum_exp)
  
  x |> 
  dplyr::mutate(sum_classification = dplyr::if_else(sum_exp > 0, "Tumor", "Normal"))})

for(i in 1:length(marker_sum_exp)) {
  p <- ggplot(marker_sum_exp[[i]], aes(x = UMAP1, y = UMAP2, color = sum_classification)) +
  geom_point(size = 0.5, alpha = 1) +
  ggtitle(names(umap_df_list_excluded_scaled)[i])
  print(p)

  #save plots
  ggsave(
    filename = paste0(
      module_base,
<<<<<<< HEAD
      '/plots/UMAP_classification_cluster_',
=======
      '../plots/UMAP_classification_cluster_',

> >>>>>>> danhtruong/danhtruong/cell-type-dsrct

Reviewer: Danh: there are a number of places where it seems like your latest changes have introduced merge artifacts like this one. Can you please fix these and rerender the notebook? I would expect to see a newer version of the html that matches the latest updates.

Reviewer: Apologies... on closer inspection it looks like you did rerun the notebook, but then a later merge brought in the above error (which should have been caught by our pre-commit tests, but wasn't for some reason!). I have reverted the offending commit and merged in the changes from the `main` repository, so this should be ready to go, but you will need to pull down the changes from this branch (or start a new branch from `main` after this is merged in) before proceeding with your next steps.
